lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,351 @@
+  }
+  if (getComputedStyle(parent).overflowY !== 'visible') {
+    const parentBCR = parent.getBoundingClientRect();
+    clientRects = clientRects.filter(cr => rectContains(parentBCR, cr));
+  }
+  if (parent.parentElement && parent.parentElement.tagName !== 'BODY') {
+    return filterClientRectsWithinAncestorsVisibleScrollArea(
+      parent,
+      clientRects
+    );
+  }
+  return clientRects;
+}
+
+/**
+ * @param {Element} node
+ * @param {boolean} includeChildren
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function getClientRects(node, includeChildren = true) {
+  /** @type {LH.Artifacts.Rect[]} */
+  let clientRects = Array.from(
+    node.getClientRects()
+  ).map(clientRect => {
+    // Contents of DOMRect get lost when returned from Runtime.evaluate call,
+    // so we convert them to plain objects.
+    const {width, height, left, top, right, bottom} = clientRect;
+    return {width, height, left, top, right, bottom};
+  });
+  if (includeChildren) {
+    for (const child of node.children) {
+      clientRects = clientRects.concat(getClientRects(child));
+    }
+  }
+
+  return clientRects;
+}
+
+/**
+ * @param {Element} node
+ * @returns {boolean}
+ */
+/* istanbul ignore next */
+function nodeHasParentTapTarget(node) {
+  if (!node.parentElement) {
+    return false;
+  }
+  if (node.parentElement.matches(tapTargetsSelector)) {
+    return true;
+  }
+  return nodeHasParentTapTarget(node.parentElement);
+}
+
+/**
+ * Check if node is in a block of text, such as paragraph with a bunch of links in it.
+ * Makes a reasonable guess, but for example gets it wrong if the element is surounded by other
+ * HTML elements instead of direct text nodes.
+ * @param {Node} node
+ * @returns {boolean}
+ */
+/* istanbul ignore next */
+function nodeIsInTextBlock(node) {
+  /**
+   * @param {Node} node
+   * @returns {boolean}
+   */
+  function isInline(node) {
+    if (node.nodeType === Node.TEXT_NODE) {
+      return true;
+    }
+    if (node.nodeType !== Node.ELEMENT_NODE) {
+      return false;
+    }
+    const {display} = getComputedStyle(/** @type {Element} */ (node));
+    return display === 'inline' || display === 'inline-block';
+  }
+
+  /**
+   * @param {Node} node
+   */
+  function hasTextNodeSiblingsFormingTextBlock(node) {
+    if (!node.parentElement) {
+      return false;
+    }
+
+    const parentElement = node.parentElement;
+
+    const nodeText = node.textContent || '';
+    const parentText = parentElement.textContent || '';
+    if (parentText.length - nodeText.length < 5) {
+      // Parent text mostly consists of this node, so the parent
+      // is not a text block container
+      return false;
+    }
+
+    for (const sibling of node.parentElement.childNodes) {
+      if (sibling === node) {
+        continue;
+      }
+      const siblingTextContent = (sibling.textContent || '').trim();
+      if (sibling.nodeType === Node.TEXT_NODE && siblingTextContent.length > 0) {
```

Comment:
why doesn't the `parentElement.textContent` check above catch this case? Shouldn't all the text of any siblings be in there?

Reply:
Yes, but parent text content also contains the content of siblings that are elements rather than text nodes.

This should not be detected as a text block:

```
<div>
   <a onClick=()>Do something</a>
   <a onClick=()>Do more stuff</a>
</div>
```

This should be counted as a text block:

```
<div>
   <a>Lighthouse</a> is a tool for web developers.
</div>
```

This should be ideally be detected as a text block but it isn't, and we're ok with that:

```
<div>
   <a>Lighthouse</a> <span>is a tool for web developers.</span>
</div>
```

Will add a comment to the code.

Reply:
> Will add a comment to the code.

great, that definitely helps